File: scripts/generate-stories.ts
import fs from 'fs';
import path from 'path';

const UI_DIR = path.resolve(process.cwd(), 'registry/web-app/ui');

// Walk directory to find all .tsx files
function walk(dir: string, fileList: string[] = []) {
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const filePath = path.join(dir, file);
    if (fs.statSync(filePath).isDirectory()) {
      walk(filePath, fileList);
    } else if (filePath.endsWith('.tsx') && !filePath.endsWith('.test.tsx') && !filePath.endsWith('.stories.tsx')) {
      fileList.push(filePath);
    }
  }
  return fileList;
}

const componentRegex = /export (?:const|function|class) ([A-Z][a-zA-Z0-9_]+)/g;
const exportListRegex = /export\s*{\s*([^}]+)\s*}/g;

function generateStory(filePath: string) {
  const content = fs.readFileSync(filePath, 'utf-8');
  let match;
  const components: string[] = [];

  while ((match = componentRegex.exec(content)) !== null) {
    components.push(match[1]);
  }

  while ((match = exportListRegex.exec(content)) !== null) {
    const exports = match[1].split(',').map(s => s.trim().split(' ')[0]);
    for (const exp of exports) {
      if (exp.match(/^[A-Z][a-zA-Z0-9_]+$/) && !components.includes(exp)) {
        components.push(exp);
      }
    }
  }

  if (components.length === 0) return;

  const mainComponent = components[0];
  const storyFilePath = filePath.replace(/\.tsx$/, '.stories.tsx');

  if (fs.existsSync(storyFilePath)) {
    console.log(`Story already exists: ${storyFilePath}`);
    return;
  }

  const relativePath = path.basename(filePath, '.tsx');
  const dirName = path.basename(path.dirname(filePath));
  const isNested = dirName !== 'ui';
  const groupName = isNested ? `Web App/${dirName}` : `Web App/${mainComponent}`;
  
  const storyContent = `import type { Meta, StoryObj } from '@storybook/react';
import { ${components.join(', ')} } from './${relativePath}';

const meta: Meta<typeof ${mainComponent}> = {
  title: '${groupName}/${mainComponent}',
  component: ${mainComponent},
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof ${mainComponent}>;

export const Default: Story = {
  args: {
    // Add default args here
  },
};
`;

  fs.writeFileSync(storyFilePath, storyContent);
  console.log(`Generated story: ${storyFilePath}`);
}

const tsxFiles = walk(UI_DIR);
for (const file of tsxFiles) {
  generateStory(file);
}

File: scripts/generate-stories.py
import os
import re

REG_BASE = '/mnt/D/Work/MVL/mvl-ui-registry/registry/web-app/ui'

def get_component_name(filename_or_dirname):
    # Convert kebab-case or snake_case to PascalCase
    parts = re.split(r'[-_]', filename_or_dirname)
    return ''.join(p.capitalize() for p in parts)

def generate_story(component_name, file_path, story_path, is_folder=False):
    # Determine the import path
    if is_folder:
        import_path = f'./{os.path.basename(file_path)}'
    else:
        import_path = f'./{os.path.basename(file_path).replace(".tsx", "")}'
        
    template = f"""import type {{ Meta, StoryObj }} from '@storybook/react';
import {{ {component_name} }} from '{import_path}';

const meta = {{
  title: 'web-app/ui/{component_name}',
  component: {component_name} as any,
  tags: ['autodocs'],
  argTypes: {{}},
}} satisfies Meta<typeof {component_name}>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {{
  args: {{}},
}};
"""
    with open(story_path, 'w', encoding='utf-8') as f:
        f.write(template)

def main():
    items = os.listdir(REG_BASE)
    for item in items:
        full_path = os.path.join(REG_BASE, item)
        
        # Skip if it's already a story file
        if item.endswith('.stories.tsx') or item.endswith('.test.tsx') or item == 'index.ts':
            continue
            
        # For a single file like dialog.tsx
        if os.path.isfile(full_path) and item.endswith('.tsx'):
            base_name = item.replace('.tsx', '')
            story_file = os.path.join(REG_BASE, f"{base_name}.stories.tsx")
            if not os.path.exists(story_file):
                comp_name = get_component_name(base_name)
                print(f"Generating story for {comp_name} ({story_file})")
                generate_story(comp_name, full_path, story_file, is_folder=False)
                
        # For a folder like checkbox/
        elif os.path.isdir(full_path):
            base_name = item
            story_file = os.path.join(REG_BASE, f"{base_name}.stories.tsx")
            if not os.path.exists(story_file):
                comp_name = get_component_name(base_name)
                print(f"Generating story for {comp_name} ({story_file})")
                generate_story(comp_name, full_path, story_file, is_folder=True)

if __name__ == '__main__':
    main()
